### ISSUE TITLE
compiler object methods don't warn about unknown kwargs

### ISSUE THREAD

**Issue by ystreet** [mesonbuild/meson#2364]
As mentioned on the tin.

https://cgit.freedesktop.org/gstreamer/gst-plugins-bad/tree/gst-libs/gst/gl/meson.build?id=980ddfdfb740107ce8d9db3f8679d235af757412#n366

Which results in:

```
Command line:  ccache cc /tmp/tmp3s718w9b/testfile.c -O0 -pipe -D_FILE_OFFSET_BITS=64 -c -o /tmp/tmp3s718w9b/output.obj 

Code:
 void f (void) {}
Compiler stdout:
 
Compiler stderr:
```

We either need to add support for the prefix kwarg or error out on its use in cc.compiles().

GStreamer bug: https://bugzilla.gnome.org/show_bug.cgi?id=787964

**Comment by nirbheek** [comment]
Hmm, looks like we don't use the `@permittedKwargs` decorator for compiler methods.

**Comment by tp-m** [comment]
Is there a consensus on whether `cc.compiles()` should support `prefix:` or not?

**Comment by nirbheek** [comment]
No, but I guess we should add it just for consistency? It would make the reference manual simpler too.

**Comment by tp-m** [comment]
We should decide if we want a `prefix:` kwarg for `cc.compiles()` or not.

If not, we can close this bug, because we do get warnings now. If yes, we should retitle.

**Comment by bonzini** [comment]
Should be fixed by typed_kwargs at this point

**Comment by eli-schwartz** [comment]
Yup -- and with regard to the general question of "should cc.compiles() support prefix" I think not. The downstream fix here was to mechanically change:

```meson
cc.compiles(mycode, prefix: myprefix)
cc.compiles('int main() { do_thing(); return 0; }', prefix: '#include <foo.h>')
```

to
```meson
cc.compiles(myprefix + '\n' + mycode)
cc.compiles('''
#include <foo.h>

int main() { do_thing(); return 0; }
''')
```

and I don't think that splitting this into separate kwargs helps readability.
